### ISSUE TITLE
Memory leak in localization mode?

### ISSUE THREAD

**Issue by matlabbe** [introlab/rtabmap#494]
How to reproduce the problem:
 1. Create a database using the [robot mapping demo](http://wiki.ros.org/rtabmap_ros#Robot_mapping). There should be around 228 nodes added to map at the end of the bag and the System Monitor would show around 430 MB usage for rtabmap node.

 2. Using valgrind's [massif](http://valgrind.org/docs/manual/ms-manual.html) tool, reprocess the database offline:
    ````bash
    $ valgrind --tool=massif --time-unit=ms --detailed-freq=1 --max-snapshots=1000 rtabmap-reprocess rtabmap.db output.db
    ````
 
 3. Opening the resulting massif log in [massif-visualizer](https://github.com/KDE/massif-visualizer), we get (memory is increasing as expected in SLAM mode):
    ![massif_demo_robot_mapping](https://user-images.githubusercontent.com/2319645/72463778-09640180-37a2-11ea-9d17-b7de6183ce4d.png)

 4. Now, redo massif in localization mode:
    ````bash
    $ valgrind --tool=massif --time-unit=ms --detailed-freq=1 --max-snapshots=1000 rtabmap-reprocess --Mem/IncrementalMemory false "rtabmap.db;rtabmap.db" output.db
    ````

 5. See the resulting result:
    ![massif_demo_robot_localization](https://user-images.githubusercontent.com/2319645/72463891-43350800-37a2-11ea-86b3-6a456550c241.png)

The question is why after initialization at the beginning, the memory is still increasing? It seems everything is constant, except the bottom red area (the colored labels represent the peaks, most of them happening at the beginning).

**Comment by timonmeyer** [comment]
Hey @matlabbe,
Is there any update regarding this issue?

**Comment by matlabbe** [comment]
I remember having done some memory optimizations and added statistics past in September/October. The results I had didn't show leaks on the experiments I've done though, so it is why I didn't update this issue. 

For this example, I think I've found what is happening. The memory is increasing because proximity detection with laser scans is enabled. On proximity detection, scans of the map are decompressed and **kept not compressed in memory afterwards**. It is also why the final memory usage is mostly the same after the run.

https://github.com/introlab/rtabmap/blob/481a140f845a7488aad9b2805b3997989ad1b6ef/corelib/src/Memory.cpp#L3074-L3080

SLAM ( max = 344.6 MB):
```
valgrind --tool=massif  rtabmap-reprocess rtabmap.db output.db
```
![Screenshot from 2021-02-10 18-20-50](https://user-images.githubusercontent.com/2319645/107589003-015bd880-6bd3-11eb-81eb-26c1bc3e5ddc.jpg)

Localization (with proximity detection enabled, max = 344.6 MB ->exactly the same size than at the end of SLAM phase = **No leaks**):
```
valgrind --tool=massif  rtabmap-reprocess --Mem/IncrementalMemory false "rtabmap.db;rtabmap.db" output.db
```
![Screenshot from 2021-02-10 18-16-23](https://user-images.githubusercontent.com/2319645/107589011-0587f600-6bd3-11eb-9852-942bfe00fae4.jpg)

To compare, here is localization without proximity detection and without loop closure, to avoid uncompressing the scans. The orange (some sqlite3 calls) is increasing, but that memory is related to DBReader::getNextData() used to load the data to be published for localization.
```
valgrind --tool=massif  rtabmap-reprocess --Mem/IncrementalMemory false --RGBD/ProximityBySpace false --Kp/MaxFeatures -1 "rtabmap.db;rtabmap.db" output.db
```
![Screenshot from 2021-02-10 18-15-58](https://user-images.githubusercontent.com/2319645/107589017-0ae54080-6bd3-11eb-8490-6f89b8234f49.jpg)

Localization (without proximity detection but with loop closure enabled without laser scans). This setup is like doing only visual loop closure detection (in a setup without scans).
```
valgrind --tool=massif  rtabmap-reprocess --Mem/IncrementalMemory false --RGBD/ProximityBySpace false --Reg/Strategy 0 "rtabmap.db;rtabmap.db" output.db
```
![Screenshot from 2021-02-10 18-24-41](https://user-images.githubusercontent.com/2319645/107589025-10db2180-6bd3-11eb-9264-155324dd043d.jpg)

In conclusion, I am not sure why DBReader::getNextData() increases for some time and stop increasing after a while. There is no leak on rtabmap side though.

**Comment by timonmeyer** [comment]
Thank you for the very detailed explanation!!

**Comment by ksvbka** [comment]
Thank for very interesting topic. I can learn alot from it!
